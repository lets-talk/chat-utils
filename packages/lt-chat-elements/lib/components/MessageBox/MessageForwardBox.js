import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

// Icons to be used on this component
import FaForward from 'react-icons/lib/fa/mail-forward';

const StyledMessageForwardBox = styled.div`
  display: flex;
  width: 30px;
  height: 30px;
  border-radius: 20px;
  background: #fff;
  position: absolute;
  flex-direction: row;
  align-self: center;
  align-items: center;
  justify-content: center;
  box-shadow: 0 0 5px 0 rgba(164, 164, 164, 1);
  cursor: pointer;
  transition: all 0.3s ease;
  top: 0;
  bottom: 0;
  margin: auto;

  ${(props) =>
    props.position === 'left' &&
      css`
        right: -50px
      `};

  ${(props) =>
    props.position === 'right' &&
      css`
        left: -50px
      `};
`;

const MessageForwardBox = (props) => (
  <StyledMessageForwardBox
    position={props.position}
    onClick={props.onForwardClick}
    onKeyPress={props.onForwardClick}
  >
    <FaForward color="#000" />
  </StyledMessageForwardBox>
);

MessageForwardBox.propTypes = {
  /**
   * Position: What side the forward is displayed
   */
  position: PropTypes.string,
  /**
   * onForwardClick: Handler function to be called when user click on forwared message
   */
  onForwardClick: PropTypes.func,
};

export default MessageForwardBox;
