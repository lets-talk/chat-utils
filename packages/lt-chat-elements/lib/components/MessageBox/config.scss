@import "../../common/styles/variables";
@import "../../common/styles/themes/themify";

$messageBox-title-font-size: $font-size-caption;
$messageBox-title-font-weight: $font-weight-semi-bold;
$messageBox-title-color: $color-white;
$messageBox-time-font-size: $font-size-caption;

$messageBox-side-margin: 10px;
$messageBox-border-radius: 5px;
$messageBox-box-shadow: none;
