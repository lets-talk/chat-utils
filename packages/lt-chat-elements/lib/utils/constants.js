module.exports = Object.freeze({
  messagesTypes: {
    SYSTEM: 'System',
    TEXT: 'Text',
    TYPING: 'Typing',
    TIME: 'Time',
    ACTIONABLE: 'Actionable',
    FILE: 'File',
    PHOTO: 'Photo',
  },
  messagesStatus: {
    WAITING: 'waiting',
    SENT: 'sent',
    RECEIVED: 'received',
    READ: 'read',
  },
  personTypes: {
    AGENT: 'Agent',
    BOT: 'Bot',
    CLIENT: 'Client',
  },
});
