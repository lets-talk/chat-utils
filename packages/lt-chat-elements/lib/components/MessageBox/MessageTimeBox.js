import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import moment from 'moment';

// Icons to be used on this component
import IoDoneAll from 'react-icons/lib/io/android-done-all';
import MdIosTime from 'react-icons/lib/md/access-time';
import MdCheck from 'react-icons/lib/md/check';

// Constants
import constants from '../../utils/constants';
import { textColor } from '../../utils/style';

const { messagesTypes } = constants;

const StyledMessageTimeBoxContainer = styled.div`
  max-width: 80px;
  text-align: right;
  color: ${(props) => textColor(props.theme, 'light', 'disabled')};
  font-size: ${(props) => props.theme.typography.classes.caption.fontSize};
  line-height: ${(props) => props.theme.typography.classes.caption.lineHeight};
  font-weight: ${(props) => props.theme.typography.weights.fontWeightMedium};
  float: right
  position: absolute;
  bottom: 0;
  right: 0;
`;
StyledMessageTimeBoxContainer.displayName = 'StyledMessageTimeBoxContainer';

const StyledTimeAgoText = styled.span`
  font-size: ${(props) => props.theme.typography.classes.small.fontSize};
`;
StyledTimeAgoText.displayName = 'StyledTimeAgoText';

const StyledStatusSpan = styled.span`
  margin-left: 3px;
`;
StyledStatusSpan.displayName = 'StyledStatusSpan';

const MessageTimeBox = (props) => (
  <StyledMessageTimeBoxContainer>
    {
      props.date && (props.type !== messagesTypes.TYPING) &&
      <StyledTimeAgoText>
        {props.dateString || moment(props.date).fromNow()}
      </StyledTimeAgoText>
    }
    {
      props.status &&
        <StyledStatusSpan>
          {
            props.status === 'waiting' &&
            <MdIosTime />
          }

          {
            props.status === 'sent' &&
            <MdCheck />
          }

          {
            props.status === 'received' &&
            <IoDoneAll />
          }

          {
            props.status === 'read' &&
            <IoDoneAll color="#4FC3F7" />
          }
        </StyledStatusSpan>
    }
  </StyledMessageTimeBoxContainer>
);

MessageTimeBox.propTypes = {
  /**
   * Type of Message: Type of the message. Values are defined as constants.
   * The current supported types are: SYSTEM, TEXT, TYPING, TIME, ACTIONABLE, FILE, PHOTO.
   */
  type: PropTypes.string.isRequired,
  /**
   * status: Message status. Values are defined as constants.
   * Current supported values are: WAITING, SENT, RECEIVED, READ
   */
  status: PropTypes.string,
  /**
   * date: Message creation date.
   */
  date: PropTypes.instanceOf(Date),
  /**
   * dateString: Formated date string to show. This is the default to use.
   * If this is not provided by default to show the time moment(props.date).fromNow() is used.
   */
  dateString: PropTypes.string,
};

MessageTimeBox.defaultProps = {
  status: '',
  date: null,
  dateString: '',
};

export default MessageTimeBox;
