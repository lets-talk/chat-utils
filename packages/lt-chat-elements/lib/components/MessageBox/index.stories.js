import React from 'react';
// Storybook stuff
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, select, object } from '@storybook/addon-knobs';

import { WrapWithTheme, defaultTheme, themeOptions } from '../../utils/stories';
import withTests from '../../test-utils/storiesWithTest';
// Component to show on storybook
import MessageBox from '.';

const stories = storiesOf('MessageBox', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withTests('MessageBox', 'FileMessage', 'PhotoMessage', 'SystemMessage'));

const messages = [
  {
    position: 'left',
    forwarded: true,
    type: 'Text',
    theme: 'white',
    view: 'list',
    title: 'TalkBot',
    text: 'Buenos días! En que podemos ayudarte?',
    status: 'waiting',
    date: '2017-11-15T20:34:27.173Z',
    dateString: '18:23',
    person: {
      avatar: '',
      email: '',
      type: 'Bot',
      status: 'online',
    },
  },
  {
    position: 'right',
    forwarded: true,
    type: 'File',
    theme: 'white',
    view: 'list',
    title: 'Solicitud de nueva tarjeta Banco',
    text: 'Solicitud de nueva tarjeta Banco',
    data: {
      uri: 'https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg',
      status: {
        click: true,
        loading: 70,
      },
      width: 300,
      height: 300,
      size: '2Mb',
    },
    status: 'sent',
    date: '2017-11-15T20:34:27.173Z',
    dateString: '17:34',
    person: {
      avatar: 'https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg',
      email: '',
      type: 'Client',
      status: 'sleeping',
    },
  },
  {
    position: 'right',
    forwarded: true,
    type: 'Photo',
    theme: 'white',
    view: 'list',
    title: 'Solicitud de nueva tarjeta Banco',
    text: 'Solicitud de nueva tarjeta Banco',
    data: {
      uri: 'http://i46.tinypic.com/sexbb8.pngg',
      status: {
        click: true,
        loading: 70,
      },
      width: 150,
      height: 150,
      size: '2Mb',
    },
    status: 'sent',
    date: '2017-11-15T20:34:27.173Z',
    dateString: '17:34',
    person: {
      avatar: 'https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg',
      email: '',
      type: 'Client',
      status: 'online',
    },
  },
  {
    position: 'left',
    forwarded: false,
    type: 'Actionable',
    theme: 'white',
    view: 'list',
    text: 'Datos Contacto',
    data: {
      actions: [
        {
          id: 1,
          name: 'Emergencia Bancaria',
        },
        {
          id: 2,
          name: 'Tarjeta de crédito',
        },
        {
          id: 3,
          name: 'Crédito Personal',
        },
        {
          id: 4,
          name: 'Otro',
        },
      ],
    },
    status: 'waiting',
    date: '2017-11-15T20:34:27.173Z',
    dateString: '18:23',
    person: {
      avatar: 'https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg',
      email: '',
      type: 'Client',
      status: 'sleeping',
    },
  },
  {
    position: 'right',
    forwarded: false,
    type: 'Typing',
    theme: 'white',
    view: 'list',
    title: '',
    text: '',
    status: 'read',
    date: '2017-11-14T18:25:22.133Z',
    dateString: '18:25',
    person: {
      avatar: 'https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg',
      email: '',
      type: 'Agent',
      status: 'offline',
    },
  },
  {
    position: 'left',
    forwarded: false,
    type: 'Time',
    theme: 'white',
    view: 'list',
    title: 'TalkBot',
    text: '15/11/2017',
    status: 'waiting',
    date: '2017-11-15T20:34:27.173Z',
    dateString: '18:23',
    person: {
      avatar: '',
      email: '',
      type: 'Bot',
      status: 'online',
    },
  },
  {
    position: 'left',
    forwarded: false,
    type: 'System',
    theme: 'white',
    view: 'list',
    title: 'TalkBot',
    text: 'Sandino ha agregado a Cristian a la conversación',
    status: 'waiting',
    date: '2017-11-15T20:34:27.173Z',
    dateString: '18:23',
    person: {
      avatar: '',
      email: '',
      type: 'Bot',
      status: 'online',
    },
  },
];

stories.addWithInfo(
  'MessageBox - Text Message',
  'TextMessage type use',
  () => (
    <div style={{ position: 'fixed', bottom: 0, right: '10px' }}>
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <MessageBox
          message={object('message', messages[0])}
          onOpen={action('onOpen')}
          onDownload={action('onDownload')}
          onTitleClick={action('onTitleClick')}
          onForwardClick={action('onForwardClick')}
          onMessageClick={action('onMessageClick')}
          onActionMessageClick={action('onActionMessageClick')}
        />
      </WrapWithTheme>
    </div>
  )
);

stories.addWithInfo(
  'MessageBox - File Message Type',
  'FileMessage type use',
  () => (
    <div style={{ position: 'fixed', bottom: 0, right: '10px' }}>
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <MessageBox
          message={object('message', messages[1])}
          onOpen={action('onOpen')}
          onDownload={action('onDownload')}
          onTitleClick={action('onTitleClick')}
          onForwardClick={action('onForwardClick')}
          onMessageClick={action('onMessageClick')}
          onActionMessageClick={action('onActionMessageClick')}
        />
      </WrapWithTheme>
    </div>
  )
);

stories.addWithInfo(
  'MessageBox - Photo Message Type',
  'Photo type use',
  () => (
    <div style={{ position: 'fixed', bottom: 0, right: '10px' }}>
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <MessageBox
          message={object('message', messages[2])}
          onOpen={action('onOpen')}
          onDownload={action('onDownload')}
          onTitleClick={action('onTitleClick')}
          onForwardClick={action('onForwardClick')}
          onMessageClick={action('onMessageClick')}
          onActionMessageClick={action('onActionMessageClick')}
        />
      </WrapWithTheme>
    </div>
  )
);

stories.addWithInfo(
  'MessageBox - Actionable Message Type',
  'Message of ActionableMessageBox type',
  () => (
    <div style={{ position: 'fixed', bottom: 0, right: '10px' }}>
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <MessageBox
          message={object('message', messages[3])}
          onOpen={action('onOpen')}
          onDownload={action('onDownload')}
          onTitleClick={action('onTitleClick')}
          onForwardClick={action('onForwardClick')}
          onMessageClick={action('onMessageClick')}
          onActionMessageClick={action('onActionMessageClick')}
        />
      </WrapWithTheme>
    </div>
  )
);

stories.addWithInfo(
  'MessageBox - Typing Message Type',
  'Message of Typing type',
  () => (
    <div style={{ position: 'fixed', bottom: 0, right: '10px' }}>
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <MessageBox
          message={object('message', messages[4])}
          onOpen={action('onOpen')}
          onDownload={action('onDownload')}
          onTitleClick={action('onTitleClick')}
          onForwardClick={action('onForwardClick')}
          onMessageClick={action('onMessageClick')}
          onActionMessageClick={action('onActionMessageClick')}
        />
      </WrapWithTheme>
    </div>
  )
);

stories.addWithInfo(
  'MessageBox - Date Message Type',
  'Message of Date type',
  () => (
    <div style={{ position: 'fixed', bottom: 0, right: '10px' }}>
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <MessageBox
          message={object('message', messages[5])}
          onOpen={action('onOpen')}
          onDownload={action('onDownload')}
          onTitleClick={action('onTitleClick')}
          onForwardClick={action('onForwardClick')}
          onMessageClick={action('onMessageClick')}
          onActionMessageClick={action('onActionMessageClick')}
        />
      </WrapWithTheme>
    </div>
  )
);

stories.addWithInfo(
  'MessageBox - System Message Type',
  'Message of System type',
  () => (
    <div
      style={{
        position: 'fixed', bottom: 0, right: '10px', width: '100%',
      }}
    >
      <WrapWithTheme
        themeName={select('Theme', themeOptions, defaultTheme)}
      >
        <MessageBox
          message={object('message', messages[6])}
          onOpen={action('onOpen')}
          onDownload={action('onDownload')}
          onTitleClick={action('onTitleClick')}
          onForwardClick={action('onForwardClick')}
          onMessageClick={action('onMessageClick')}
          onActionMessageClick={action('onActionMessageClick')}
        />
      </WrapWithTheme>
    </div>
  )
);
