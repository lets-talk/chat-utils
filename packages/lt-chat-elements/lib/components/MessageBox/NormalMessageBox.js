import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

// Constants
import constants from '../../utils/constants';

// Other components used by this component
import FileMessage from '../FileMessage';
import PhotoMessage from '../PhotoMessage';
import Loader from '../Loader';
import ActionableMessageBox from '../ActionableMessageBox';

// Helpers components
import MessageTimeBox from './MessageTimeBox';
import MessageForwardBox from './MessageForwardBox';

const StyledForwardDiv = styled.div`
  opacity: 0;
  visibility: hidden;
`;

const StyledMessageBody = styled.div`
  margin: 0;
  padding: 0;
  position: relative;
`;

const StyledMessageTitle = styled.div`
  margin: 0;
  margin-bottom: 8px;

  color: ${(props) => props.theme.typography.classes.caption.bold.dark.secondary.color};
  font-size: ${(props) => props.theme.typography.classes.caption.bold.dark.secondary.fontSize};
  font-weight: ${(props) => props.theme.typography.classes.caption.bold.dark.secondary.fontWeight};

  user-select: none;
  cursor: pointer;
  display: flex;
  align-items: center;

  & > span:hover {
    text-decoration: underline;
  }
`;

const StyledTextMessage = styled.div`
  display: inline-block;
  width: calc(100% - 50px);
  word-break: break-word;
`;

const StyledTypingMessage = styled.div`
  word-break: break-word;
`;

const NormalMessageBox = (props) => {
  const { messagesTypes, personTypes } = constants;

  const StyledNormalMessageContainer = styled.div`
    position: relative;
    padding-bottom: 3px;
    background-color: ${(localProps) => localProps.theme.palette.background.primary};
    border-radius: 5px;
    box-shadow: none;
    margin-top: 3px;
    flex-direction: column;
    margin-bottom: 3px;
    padding: 15px 10px;

    ${() =>
    (props.position === 'left') &&
        css`
          float: left;
          margin-left: 5px;
          margin-left: ${() => props.person.avatar ? '45px' : '0'};
        `};

    ${() =>
    (props.position === 'right') &&
        css`
          float: right;
          margin-left: 5px;
          margin-right: ${() => props.person.avatar ? '45px' : '0'};
        `};
    

    color: ${(localProps) => localProps.theme.typography.classes.footnote.normal.light.primary.color};
    font-size: ${(localProps) => localProps.theme.typography.classes.footnote.normal.light.primary.fontSize};
    font-weight: ${(localProps) => localProps.theme.typography.classes.footnote.normal.light.primary.fontWeight};
    
    ${() =>
    ((props.person.type === personTypes.BOT || props.person.type === personTypes.AGENT) && props.type !== messagesTypes.ACTIONABLE) &&
        css`
          color: ${(localProps) => localProps.theme.palette.text.primary};
          background-color: ${(localProps) => localProps.theme.palette.background.secondary};
          .letstalk-mbox-title {
            color: ${(localProps) => localProps.theme.palette.text.primary};
          }
          .letstalk-mbox-time {
            color: ${(localProps) => localProps.theme.palette.text.primary};
          }
        `};

    ${() =>
    (props.type === messagesTypes.ACTIONABLE) &&
        css`
          padding: 0;
          border-radius: 5px;
          box-shadow: 0px 0px 2px 1px rgba(94,124,139,0.3);
          background-color: ${(localProps) => localProps.theme.palette.common.white};
          min-width: 50%;
          color: ${(localProps) => localProps.theme.palette.text.primary};
        `};

    ${() =>
    (props.type !== messagesTypes.TYPING) &&
        css`
          min-width: 170px;
        `};

    &:hover ${StyledForwardDiv} {
      opacity: 1;
      visibility: visible;
    }
  `;

  return (
    <StyledNormalMessageContainer>
      <StyledMessageBody>
        {
          props.forwarded === true &&
          <StyledForwardDiv>
            <MessageForwardBox {...props} />
          </StyledForwardDiv>
        }

        {
          (props.title) &&
            <StyledMessageTitle
              onClick={props.onTitleClick}
              onKeyPress={props.onTitleClick}
            >
              {
                props.title &&
                <span>{props.title}</span>
              }
            </StyledMessageTitle>
        }

        {
          props.type === messagesTypes.TEXT &&
          <StyledTextMessage>
            {props.text}
          </StyledTextMessage>
        }

        {
          props.type === messagesTypes.TYPING &&
          <StyledTypingMessage>
            <Loader active type="ball-pulse" size="xsmall" color="rgb(113, 131, 150)" />
          </StyledTypingMessage>
        }

        {
          props.type === messagesTypes.PHOTO &&
          <PhotoMessage
            onOpen={props.onOpen}
            onDownload={props.onDownload}
            data={props.data}
            text={props.text}
          />
        }

        {
          props.type === messagesTypes.FILE &&
          <FileMessage
            onOpen={props.onOpen}
            onDownload={props.onDownload}
            data={props.data}
            text={props.text}
          />
        }

        {
          props.type === messagesTypes.ACTIONABLE &&
          <ActionableMessageBox
            data={props.data}
            onClickAction={props.onActionMessageClick}
          />
        }

        {props.type !== messagesTypes.TYPING && props.type !== messagesTypes.ACTIONABLE &&
        <MessageTimeBox
          type={props.type}
          status={props.status}
          dateString={props.dateString}
          date={props.date}
        />
        }

      </StyledMessageBody>
    </StyledNormalMessageContainer>
  );
};


NormalMessageBox.propTypes = {
  /**
   * Position: What side the message is displayed inside the chatbox
   */
  position: PropTypes.string,
  /**
   * Type of Message: Type of the message. Values are defined as constants.
   * The current supported types are: SYSTEM, TEXT, TYPING, TIME, ACTIONABLE, FILE, PHOTO.
   */
  type: PropTypes.string,
  /**
   * Text: Actual message content text.
   */
  text: PropTypes.string,
  /**
   * Title: Message title. Is used to display as a first line. It is also possible to make it clickable
   */
  title: PropTypes.string,
  /**
   * onTitleClick: Handler function to be called when user click on message title
   */
  onTitleClick: PropTypes.func,
  /**
   * onForwardClick: Handler function to be called when user click on forwared message
   */
  onForwardClick: PropTypes.func,
  /**
   * date: Message creation date.
   */
  date: PropTypes.instanceOf(Date),
  /**
   * data: Object with extra data used to display information about the message
   */
  data: PropTypes.object,
  /**
   * onClick: Handler function that is called on message click action
   */
  onClick: PropTypes.func,
  /**
   * onActionMessageClick: Handler function that is called when user click on an item of
   * a message that contains actionable action list
   */
  onActionMessageClick: PropTypes.func,
  /**
   * onOpen: Handler function that is called when user tries to open a File/Photo message type
   */
  onOpen: PropTypes.func,
  /**
  * onDownload: Handler function that is called when user tries to download a File/Photo message type
  */
  onDownload: PropTypes.func,
  /**
  * forwarded: Boolean that indicates if message was forwareded. If this is false (default) onForwardClick
  * handler has no effect at all
  */
  forwarded: PropTypes.bool,
  /**
   * status: Message status. Values are defined as constants.
   * Current supported values are: WAITING, SENT, RECEIVED, READ
   */
  status: PropTypes.string,
  /**
   * dateString: Formated date string to show. This is the default to use.
   * If this is not provided by default to show the time moment(props.date).fromNow() is used.
   */
  dateString: PropTypes.string,
  /**
   * person: Object representing the person that created or submited this message.
   */
  person: PropTypes.object,
  /**
   * className: Extra className to provide to the component in order to style it when used in different contexts.
   */
  className: PropTypes.string,
};

NormalMessageBox.defaultProps = {
  position: 'left',
  type: 'text',
  text: '',
  title: null,
  onTitleClick: null,
  onActionMessageClick: null,
  onForwardClick: null,
  date: new Date(),
  data: {},
  onClick: null,
  onOpen: null,
  onDownload: null,
  forwarded: false,
  status: '',
  dateString: null,
  person: null,
  className: '',
};


export default NormalMessageBox;
