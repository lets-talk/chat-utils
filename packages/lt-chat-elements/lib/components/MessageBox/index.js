import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

// Constants
import constants from '../../utils/constants';

// Types
import { MessagePropType } from '../../utils/types';

// Other components used by this component
import SystemMessage from '../SystemMessage';
import TimeMarkMessage from '../TimeMarkMessage';
import Avatar from '../Avatar';

// Helpers components
import NormalMessageBox from './NormalMessageBox';

const StyledMessageBoxContainer = styled.div`
  position: relative;
  flex-direction: column;
  display: block;
  overflow: hidden;
  min-width: 200px;
  &:hover div {
    opacity: 1;
    visibility: visible;
  }
`;

const StyledAvatarContainer = styled.div`
  display: inline-block;
  position: absolute;
  bottom: 0;
  ${(props) =>
    props.position === 'left' &&
      css`
        float: left;
        left: 0;
      `};

  ${(props) =>
    props.position === 'right' &&
      css`
        float: right;
        right: 0;
      `};
`;

const { messagesTypes } = constants;

const renderAutomaticMessage = (type, text) => {
  if (type === messagesTypes.SYSTEM) {
    return (<SystemMessage text={text} />);
  } else if (type === messagesTypes.TIME) {
    return (<TimeMarkMessage text={text} />);
  }
  return <div />;
};
const renderNormalMessage = (props) => <NormalMessageBox {...props} />;

const MessageBox = (props) => {
  const { className, onMessageClick, message } = props;
  const {
    type, text, person, position,
  } = message;

  return (
    <StyledMessageBoxContainer
      className={className}
      onClick={onMessageClick}
      onKeyPress={onMessageClick}
    >
      {
        person.avatar && type !== messagesTypes.SYSTEM &&
        <StyledAvatarContainer position={position}>
          <Avatar
            src={person.avatar}
            withStatus={false}
            size="xsmall"
          />
        </StyledAvatarContainer>
      }

      {
        type === messagesTypes.SYSTEM || type === messagesTypes.TIME ?
          renderAutomaticMessage(type, text)
          :
          renderNormalMessage(props)
      }
    </StyledMessageBoxContainer>
  );
};


MessageBox.propTypes = {
  /**
   * message: Actual message object.
   */
  message: PropTypes.shape(MessagePropType).isRequired,
  /**
   * onMessageClick: Handler function that is called on message click action
   */
  onMessageClick: PropTypes.func,
  /**
   * className: Extra className to provide to the component in order to style it when used in different contexts.
   */
  className: PropTypes.string,
};

MessageBox.defaultProps = {
  position: 'left',
  type: 'text',
  text: '',
  onMessageClick: null,
  person: null,
  className: '',
};


export default MessageBox;
