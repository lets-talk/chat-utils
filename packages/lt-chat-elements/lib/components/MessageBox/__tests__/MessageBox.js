import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { mountWithTheme } from '../../../test-utils';
// Constants
import constants from '../../../utils/constants';
import MessageBox from '../index';

describe('MessageBox component', () => {
  const { messagesTypes } = constants;

  const person = {
    avatar: 'https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg',
    email: '',
    type: 'Client',
  };

  it('should render a System message', () => {
    const props = {
      message: {
        type: messagesTypes.SYSTEM,
        position: 'left',
        status: 'read',
        person,
      },
    };
    const component = shallow(<MessageBox {...props} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
    // Expect to render a <SystemMessage>
    expect(component.find('SystemMessage').length).toBe(1);
  });

  it('should render a Time message', () => {
    const props = {
      message: {
        type: messagesTypes.TIME,
        position: 'right',
        status: 'read',
        person,
      },
    };

    const component = shallow(<MessageBox {...props} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
    // Expect to render a <TimeMarkMessage>
    expect(component.find('TimeMarkMessage').length).toBe(1);
  });

  it('should render a Normal message', () => {
    const props = {
      message: {
        type: messagesTypes.TEXT,
        position: 'left',
        status: 'read',
        person,
      },
    };

    const component = shallow(<MessageBox {...props} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
    // Expect to render a <TimeMarkMessage>
    expect(component.find('NormalMessageBox').length).toBe(1);
  });

  it('should Mount a Normal message on the left', () => {
    const props = {
      message: {
        type: messagesTypes.TEXT,
        position: 'left',
        status: 'read',
        person,
      },
    };

    const component = mountWithTheme(<MessageBox {...props} />);

    expect(component.length).toBe(1);
    // Expect to render a <TimeMarkMessage>
    expect(component.find('NormalMessageBox').length).toBe(1);
    expect(component.find('StyledMessageBoxAvatarContainer').props().position).toBe('left');
  });

  it('should Mount a Normal message on the right', () => {
    const props = {
      message: {
        type: messagesTypes.TEXT,
        position: 'right',
        status: 'sent',
        person,
      },
    };

    const component = mountWithTheme(<MessageBox {...props} />);

    expect(component.length).toBe(1);
    // Expect to render a <TimeMarkMessage>
    expect(component.find('NormalMessageBox').length).toBe(1);
    expect(component.find('StyledMessageBoxAvatarContainer').props().position).toBe('right');
  });
});
