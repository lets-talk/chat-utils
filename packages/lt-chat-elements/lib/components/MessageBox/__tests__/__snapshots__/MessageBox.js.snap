// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBox component should render a Normal message 1`] = `
<StyledMessageBoxContainer
  className=""
  onClick={null}
  onKeyPress={null}
>
  <StyledMessageBoxAvatarContainer
    position="left"
  >
    <Avatar
      alt=""
      className="LT-Avatar-Container"
      color={null}
      size="xsmall"
      sizes={null}
      src="https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg"
      srcSet={null}
      status=""
      type="circle"
      withStatus={false}
    />
  </StyledMessageBoxAvatarContainer>
  <NormalMessageBox
    className=""
    message={
      Object {
        "person": Object {
          "avatar": "https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg",
          "email": "",
          "type": "Client",
        },
        "position": "left",
        "status": "read",
        "type": "Text",
      }
    }
    onActionMessageClick={null}
    onClick={null}
    onDownload={null}
    onForwardClick={null}
    onMessageClick={null}
    onOpen={null}
    onTitleClick={null}
    person={null}
    position="left"
    text=""
    type="text"
  />
</StyledMessageBoxContainer>
`;

exports[`MessageBox component should render a System message 1`] = `
<StyledMessageBoxContainer
  className=""
  onClick={null}
  onKeyPress={null}
>
  <SystemMessage
    className=""
    message={
      Object {
        "person": Object {
          "avatar": "https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg",
          "email": "",
          "type": "Client",
        },
        "position": "left",
        "status": "read",
        "type": "System",
      }
    }
    onMessageClick={null}
    person={null}
    position="left"
    text=""
    type="text"
  />
</StyledMessageBoxContainer>
`;

exports[`MessageBox component should render a Time message 1`] = `
<StyledMessageBoxContainer
  className=""
  onClick={null}
  onKeyPress={null}
>
  <StyledMessageBoxAvatarContainer
    position="right"
  >
    <Avatar
      alt=""
      className="LT-Avatar-Container"
      color={null}
      size="xsmall"
      sizes={null}
      src="https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg"
      srcSet={null}
      status=""
      type="circle"
      withStatus={false}
    />
  </StyledMessageBoxAvatarContainer>
  <TimeMarkMessage
    className=""
    message={
      Object {
        "person": Object {
          "avatar": "https://pbs.twimg.com/profile_images/718588760003383296/2AG8omMO_400x400.jpg",
          "email": "",
          "type": "Client",
        },
        "position": "right",
        "status": "read",
        "type": "Time",
      }
    }
    onMessageClick={null}
    person={null}
    position="left"
    text=""
    type="text"
  />
</StyledMessageBoxContainer>
`;
